Add datepicker based on dialog pattern for issue 1046

jongund opened the pull request:
Adds date picker dialog example to the modal-dialog directory for issue #1046.

#### Preview Link
 
[Preview link for datepicker dialog](https://raw.githack.com/w3c/aria-practices/issue1046-datepicker-dialog/examples/dialog-modal/datepicker-dialog.html)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/pull/1054.html" title="Last updated on Jul 11, 2019, 12:54 PM UTC (147c666)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/1054/0ba3c1c...147c666.html" title="Last updated on Jul 11, 2019, 12:54 PM UTC (147c666)">Diff</a>

spectranaut commented:
Found a user interaction bug: If you click on the calendar icon, the dialog opens but focus remains on the calendar icon.
This is mostly odd because a user might click the date then try using left/right arrows, but you still have to type "ENTER" or "SPACE" or "DOWN ARROW" before the focus will move from the icon (even though the dialog is open). The date is also not in the tab sequence until you send on of those keys.

jongund commented:
Typically clicking on a button moves focus to the button, not somewhere else.  Since the user clicked on the button to open the dialog, mostly likely they will be using the mouse to select a date.

But it is something we should ask Matt about to see if he thinks it is something the group should discuss and review.

mcking65 commented on examples/dialog-modal/datepicker-dialog.html:
```diff
@@ -0,0 +1,771 @@
+          <button class="icon"
+            aria-label="Calendar"
+            aria-expanded="false"
+            aria-haspopup="dialog"
```
Please remove aria-haspopup. Adding it to a button makes it a menubutton, and this is a button that opens a dialog. The primary reason for the `dialog` value of haspopup is to support haspopup on the edit field in comboboxes. We have not added haspopup to the dialog pattern, and I don't expect we will.

mcking65 commented on examples/dialog-modal/datepicker-dialog.html:
```diff
@@ -0,0 +1,771 @@
+            aria-label="Calendar"
+            aria-expanded="false"
+            aria-haspopup="dialog"
+            aria-controls="id-datepicker-1"
```
Similarly, remove aria-controls since we also do not include this in the dialog pattern.

mcking65 commented on examples/dialog-modal/datepicker-dialog.html:
```diff
@@ -0,0 +1,771 @@
+
+            <span id="id-dialog-label"
+                  class="monthYear"
+                  role="status">
```
Role status doesn't work here because it is off by default. Please change this to aria-live="polite" and remove the status role. Also, since this is the dialog title, wrap in H2

mcking65 commented:
@jongund, I pushed a few changes to the introduction.

One other change that I didn't include in my review ... currently the button is named "Calendar". I think we should change it to "Choose Date".

jongund commented:
@mcking65 @spectranaut @a11ydoer I have made the changes requested yesterday and in the pull request to the datepicker dialog example so it should be ready for review and adding additional regression tests

spectranaut commented:
@mcking65 @jongund I read through all the descriptions in the tables and tested the widget accordingly. I added data-test-ids and fixed a few spelling mistakes and html errors. It mostly looks and works great! I only have a few bugs/comments:

1. Potential bug in table "Choose Date Button", row "space/return/down arrow"
When focus is on calendar button, "Down Arrow" does not open the calendar, instead the page scrolls down in firefox and chrome.

2. Clarity issue in table "Date Picker Dialog: Calendar Navigation Buttons", row: aria-controls="IDREF"
This is low priority, but the description column doesn't really make sense to me. The description column says: "Provide a reference for screen reader to navigate to the date grid."

3. Linting error:
'aria-describedby' is set on the "Choose Date" button, but there is no element with the id referenced. This is causing a linting error -- once that is fixed the CI tests will all succeed. If 'aria-describedby' should remain, perhaps there should be a rown in the 'Choose Date Button' table about it. For simplicity, perhaps for now it should be removed.

mcking65 commented:
@spectranaut wrote:

> 1. Potential bug in table "Choose Date Button", row "space/return/down arrow"
>    When focus is on calendar button, "Down Arrow" does not open the calendar, instead the page scrolls down in firefox and chrome.

Down arrow is behaving correctly. Down arrow should be removed from the documentation. It should not open the dialog. This is not a menu button. It is a standard button that opens a dialog.

I also noticed that the choose date button has aria-expanded when the dialog is open. That should be removed. And, the documentation about it should be removed.

@spectranaut wrote:
> 2. Clarity issue in table "Date Picker Dialog: Calendar Navigation Buttons", row: aria-controls="IDREF"
>    This is low priority, but the description column doesn't really make sense to me. The description column says: "Provide a reference for screen reader to navigate to the date grid."

I'm not sure we should have aria-controls on those buttons. There is nothing incorrect about it. But, there is nothing that calls for it either. I am adding this to tomorrow's meeting agenda.

@spectranaut wrote:
> 3. Linting error:
>    'aria-describedby' is set on the "Choose Date" button, but there is no element with the id referenced. This is causing a linting error -- once that is fixed the CI tests will all succeed. If 'aria-describedby' should remain, perhaps there should be a rown in the 'Choose Date Button' table about it. For simplicity, perhaps for now it should be removed.

It should be removed.

jongund commented:
@spectranaut @mcking65 

I made updates and pushed them to this branch based on the feedback in the previous message
1. Removed the `aria-describedby` from the "Choose Date" button
1. Removed the `aria-controls` from the buttons in the dialog box that control the month/year
1. Updated an IDREF in the test file to match the new ids in the html file

spectranaut commented:
@jongund I just added a commit to removed "aria-expanded", like @mcking65 requested above.

mcking65 commented on examples/dialog-modal/datepicker-dialog.html:
```diff
@@ -0,0 +1,739 @@
+            </tr>
+          </thead>
+          <tbody>
+            <tr data-test-id="test-not-required">
```
Please delete this row. We do not document implied roles.

jongund replied:
@mcking65 
Removed the implied button role

mcking65 commented on examples/dialog-modal/datepicker-dialog.html:
```diff
@@ -0,0 +1,739 @@
+              </tr>
+            </thead>
+            <tbody>
+                <tr data-test-id="test-not-required">
```
Please delete this row. We do not document implied roles.

jongund replied:
@mcking65 
Removed reference to implied button role

mcking65 commented:
Found bug where if you:
1. choose a date
2. back tab to date input from choose date button.
3. press delete causing placeholder to appear.
4. tab forward to choose date button and the aria-label still has the previous date instead of no date.

In all other cases, the date in the aria-label is getting updated correctly.

jongund commented:
@spectranaut @mcking65 
I made the following changes to the date picker dialog documentation
1. Update references to use "Choose Date" button
1. Added information on the `abbr` attribute in the  `columnheader` row for documentation of date grid in the usage column

spectranaut commented:
Caught I bug I introduced in my last commit while writing tests for this example...

spectranaut commented:
I'm going to work on two things this morning: 
- remove "key down" call from the code
- make sure that the date gets updated properly on delete, matt's bug.

mcking65 commented:
@jnurthen, thank you for the review and edits!! We're ready to ship this one.
